File: Leetcode/JAVA/#84_LargestRectangleInHistogram_Hard.java
/**
 * 84. Largest Rectangle in Histogram
 * Hard
 * Given n non-negative integers representing the histogram's bar height where the width of each bar is 1, find the area of largest rectangle in the histogram.
 * Above is a histogram where width of each bar is 1, given height = [2,1,5,6,2,3].
 * The largest rectangle is shown in the shaded area, which has area = 10 unit.
 *
 * Example:
 * Input: [2,1,5,6,2,3]
 * Output: 10
 *
 * Link: https://leetcode.com/problems/largest-rectangle-in-histogram/
 *
 * Help: 把柱状图逐渐下压，出现高度为0的时候，切割分治，记录最大矩阵面积
 */

class Solution {
    public int largestRectangleArea(int[] heights) {
        return heights.length==0 ? 0 : getMaxArea(heights, 0, heights.length-1, 0);
    }
    /**
     * [getMaxArea description]
     * @param  heights [输入数组]
     * @param  start   [开始下标]
     * @param  end     [结束下标]
     * @param  max     [上个记录的最大矩阵面积]
     * @return         [当前最大矩阵面积]
     */
    private int getMaxArea(int[] heights, int start, int end, int max) {
        //单个柱值，直接返回
        if(start==end){ return Math.max(heights[start],max);}
        //找出最小高度
        int min=heights[start];
        for(int i=start;i<=end;i++){ min = Math.min(heights[i], min);}
        //最大矩阵面积
        int cur_max=Math.max((end-start+1)*(min),max);
        //每次从底部切掉组内最小高度，以0为分界，拆组分治
        for(int i=start;i<=end;i++){
            if(heights[i]==min || i==end){
                //对每个区间找最低高度做切割，找不到则不切割
                int cur_end = heights[i]==min ? i-1 : i;
                int area = getMaxArea(heights, start, cur_end, cur_max);
                cur_max = Math.max(area,cur_max);
                start=i+1;
            }
        }
        return cur_max;
    }
}

File: Leetcode/C++/#84_LargestRectangleInHistogram_Hard.cpp
/**
 * 84. Largest Rectangle in Histogram
 * Hard
 * Given n non-negative integers representing the histogram's bar height where the width of each bar is 1, find the area of largest rectangle in the histogram.
 * Above is a histogram where width of each bar is 1, given height = [2,1,5,6,2,3].
 * The largest rectangle is shown in the shaded area, which has area = 10 unit.
 *
 * Example:
 * Input: [2,1,5,6,2,3]
 * Output: 10
 *
 * Link: https://leetcode.com/problems/largest-rectangle-in-histogram/
 *
 * 思路:
 * 用栈来模拟，遍历heights数组，如果大于栈顶元素，就push进去；否则，持续弹栈来计算从栈顶点到降序点的矩阵大小。然后将这一部分全部替换为降序点的值，即做到了整体依然是有序非降的。
 * 整个过程中，即把所有的局部最大矩阵计算过了，又在宽度范围内保留了全部的场景。
 * 举例，2，1，5，6，3的场景。
 * 先加入一个0，方便最后可以全部弹栈出来。变成：2，1，5，6，3，0.
 * 2进栈，1比栈顶小，对2进行出栈，area = 2；
 * 2被替换为1进栈，1继续进栈，这是栈为1，1；
 * 5，6都是非降的，继续进栈，栈为1，1，5，6；
 * 遇到3，是一个降序点；开始弹栈，6出栈，对应area=61； 5出栈对应area=52；下一个1比3小，不需要弹栈。然后将5、6的弹栈后的空位压栈为3，这是栈为1，1，3，3，3；
 * 下一步遇到0，开始依次出栈，得到area=31，32，33，14，1*5。
 * 遍历结束。整个过程中max的area为10.
 *
 * 在整个过程中，将每一个点作为起点，到各个波峰最大的几个矩阵都计算过一次，即保证了每个点作为起点的最大矩阵可能性都试探过。
 */

class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        int max_area = 0;
        heights.push_back(0);
        int sz = heights.size();
        int stack[sz];
        stack[0] = heights[0];
        int stack_idx = 0;
        int i = 1;
        while (stack_idx >= 0 && i < sz) {
            if (heights[i] >= stack[stack_idx]) {
                stack[++stack_idx] = heights[i++];
                continue;
            }
            while (stack_idx >= 0 && stack[stack_idx] > heights[i]) {
                int area = stack[stack_idx] * (i - stack_idx);
                if (area > max_area) { max_area = area; }
                stack_idx--;
            }
            while (stack_idx < i) {
                stack[++stack_idx] = heights[i];
            }
            i++;
        }
        return max_area;
    }
};